### ISSUE TITLE
Generic type is being ignored 

### ISSUE THREAD

**Issue by pleerock** [Microsoft/TypeScript#26639]
```ts
export interface Category {
    id: number;
    name: string;
    posts?: Post[];
}

export interface Post {
    id: number;
    title: string;
    category: Category;
}

export interface Photo {
    id: number;
    filename: string;
}

export class Model<ModelType> {
    constructor(name: string) {
    }
}

export const PostModel = new Model<Post>("Post");
export const CategoryModel = new Model<Category>("Category");
export const PhotoModel = new Model<Photo>("Photo");

export type ResolverOptions<ModelType> = {
    [P in keyof ModelType]?: { type: Model<ModelType[P]> }
}

const options: ResolverOptions<Post> = {
    category: {
        type: PostModel
    }
}
```

Last line (`type: PostModel`) should give an error.

It does know what type to expect right:

<img width="758" alt="screenshot 2018-08-23 23 29 04" src="https://user-images.githubusercontent.com/1753397/44544634-34246480-a72c-11e8-80c7-7135360a5e04.png">

However it accepts any instance of Model - PostModel, CategoryModel, PhotoModel without any error. Why?

<img width="325" alt="screenshot 2018-08-23 23 30 38" src="https://user-images.githubusercontent.com/1753397/44544721-7188f200-a72c-11e8-8574-8fe44ab8c6a3.png">

p.s. sorry for the bad title, hope somebody can help to name this issue properly

**Comment by andy-ms** [comment]
In `Model<ModelType>`, `ModelType` is unused. So `Model<number>` is assignable to `Model<string>` because there's no place for them to conflict. A quick fix is to add a declaration like `__nominal!: ModelType;` to `Model`.

**Comment by pleerock** [comment]
wow this is so non obvious, thanks for the answer

**Comment by pleerock** [comment]
Don't want to spam issues, but is it treated as bug?

I have following error:

<img width="534" alt="screenshot 2018-08-23 23 54 20" src="https://user-images.githubusercontent.com/1753397/44545906-bd896600-a72f-11e8-88e8-c0ab05fe6a8f.png">

But it does not provide a suggestion

<img width="431" alt="screenshot 2018-08-23 23 55 09" src="https://user-images.githubusercontent.com/1753397/44545954-e4479c80-a72f-11e8-9974-a206faaada26.png">

**Comment by andy-ms** [comment]
@pleerock Could you post a small complete example as plaintext that I could use to verify?

**Comment by pleerock** [comment]
@andy-ms there you are:

```ts
export class Model<ModelType> {
    __nominal!: ModelType;
    constructor(name: string) {
    }
}

export interface Category {
    id: number;
    name: string;
    posts?: Post[];
}

export interface Post {
    id: number;
    title: string;
    categories: Category;
}

export const PostModel = new Model<Post>("Post");
export const CategoryModel = new Model<Category>("Category");

export type ResolverOptions<ModelType> = {
    [P in keyof ModelType]?: { type: Model<ModelType[P]> }
}

export function resolver<ModelType>(
    model: Model<ModelType>,
    resolver: ResolverOptions<ModelType>
) {
}
export const PostResolver = resolver(PostModel, {
    categories: {
        // test autocompletion here
    }
});
```
